### PR TITLE
Implemented flash attention for smolvlm

### LINKED ISSUE #912: [Feature]: Implement Flash Attention for SmolVLM text decoder
### 🚀 Feature Description

Add Flash Attention support in the SmolVLM text model self-attention path to improve decode/prefill efficiency and reduce attention memory overhead.

The current attention implementation computes scaled dot-product attention explicitly using q.matmul(k_cache.t()), applies a causal mask, softmax, and then multiplies by v_cache. This works correctly, but it materializes the full attention matrix, which becomes expensive as context length grows.

### 📂 Feature Category

Rust Core Library

### 💡 Motivation

SmolVLM text decoding currently uses standard attention over the KV cache. For longer prompts or generation sequences, this can increase latency and memory usage because the attention scores are materialized before softmax.

Flash Attention would make the attention computation more memory-efficient and potentially improve end-to-end generation throughput, especially during prefill and longer-context decoding.

### 💭 Proposed Solution

Implement a Flash Attention-backed attention path inside the SmolVLM text model attention module.

The implementation should:

Preserve the existing KV cache behavior.
Support causal masking correctly for prefill.
Preserve dtype handling around the attention computation.
Fall back to the current standard attention path if Flash Attention is unavailable or unsupported on the current backend/device.
Include benchmarks comparing standard attention vs Flash Attention for prefill and decode workloads.
The relevant TODO currently exists in the attention forward path before the explicit attention computation.

### 📚 Library Reference

candle

### 🔄 Alternatives Considered

Keep the current explicit attention implementation. This is simpler and easier to debug, but it materializes the full attention matrix and may not scale well for longer contexts.

### 🎯 Use Cases

Faster SmolVLM text decoding.
Lower memory usage during long-context inference.
Better end-to-end VLM generation latency.
Cleaner benchmarking of attention backends in Kornia VLM.

### 📝 Additional Context

_No response_

### 🤝 Contribution Intent

- [x] I plan to submit a PR to implement this feature
- [ ] I'm requesting this feature but not planning to implement it

### PR DESCRIPTION
**Pull request by Varun-sai-500**
## 📝 Description

**Fixes/Relates to:** #912 
Adds a CUDA flash-attention path for SmolVLM text attention when running prefill with `fp16`/`bf16` weights, while keeping the regular SDPA fallback path safe for CPU/Metal/decode and non-half precision execution.

---

## 🛠️ Changes Made
- [x] Added a CUDA-gated flash-attention path for prefill attention when `attn_dtype` is `F16` or `BF16`.
- [x] Preserved the fallback attention path for CPU, Metal, single-token decode, and non-half precision dtypes.
- [x] Kept fallback SDPA compute in `F32` by default, while allowing `BF16` attention to remain in `BF16`.
- [x] Converted flash-attention inputs to the model attention dtype and converted outputs back to the original input dtype before projection.
- [x] Ensured flash attention only runs when the cache corresponds to the current prefill sequence.
---

## 🧪 How Was This Tested?
- [x] **Unit Tests:** (List new/updated tests)
- [x] **Manual Verification:** (Describe the steps you took)
- [x] **Performance/Edge Cases:** 
   Handles edge cases like on fp16, bf16, fp32

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [ ] 🟢 **No AI used.**
- [x] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist
- [x] I am assigned to the linked issue (required before PR submission)
- [x] The linked issue has been approved by a maintainer
- [x] This PR strictly implements what the linked issue describes (no scope creep)
- [x] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] (Optional) I have attached screenshots/recordings for UI changes.

---
Closes: #912

### CODE REVIEW

File: crates/kornia-vlm/src/smolvlm/text_model.rs [hunk 1]
```diff
@@ -106,49 +105,55 @@ impl Attention {
...
-            } else {
-                let mask =
-                    Self::generate_causal_mask(seq_len, self.cache.current_seq_len(), device)?;
-                att.broadcast_add(&mask)?
-            };
+        // fallback: normal SDPA — CPU, Metal, single-token decode
 
-            let att = candle_nn::ops::softmax_last_dim(&att)?;
+        let compute_dtype = match self.attn_dtype {
+            DType::BF16 => DType::BF16,
```
[the remark this answers is not in the document]
Comment (Varun-sai-500) [reply crates/kornia-vlm/src/smolvlm/text_model.rs]: fixed

File: crates/kornia-vlm/src/smolvlm/text_model.rs [hunk 2]
```diff
@@ -82,8 +82,7 @@ impl Attention {
     }
 
     fn forward(&mut self, x: &Tensor, index_pos: usize) -> Result<Tensor> {
-        let device = x.device();
-
+        let _device = x.device();
```
[the remark this answers is not in the document]
Comment (Varun-sai-500) [reply crates/kornia-vlm/src/smolvlm/text_model.rs]: don't use variable only, fixed issue

**Review by edgarriba — APPROVED** [review]

**Comment by Varun-sai-500** [comment]
@edgarriba CI IS GREEN
